arch/x86/kvm/pkvm/pkvm.c

```diff
@@ -226,6 +228,166 @@ static void pkvm_vm_destroy(int vm_handle, struct pkvm_memcache *mc)
 	donate_mem_in_memcache(mc);
 }
 
+static int attach_pkvm_vcpu_to_vm(struct pkvm_vm *pkvm_vm, struct pkvm_vcpu *pkvm_vcpu)
+{
+	struct kvm *kvm = &pkvm_vm->kvm;
+	int vcpu_handle;
+
+	pkvm_spin_lock(&pkvm_vm->lock);
+
+	if (kvm->created_vcpus == KVM_MAX_VCPUS) {
+		pkvm_spin_unlock(&pkvm_vm->lock);
+		return -EINVAL;
+	}
+	vcpu_handle = kvm->created_vcpus++;
+	pkvm_vcpu->vcpu.arch.pkvm_vcpu_handle = vcpu_handle;
+	pkvm_vm->vcpus[vcpu_handle] = pkvm_vcpu;
+
+	pkvm_spin_unlock(&pkvm_vm->lock);
+
+	atomic_set(&pkvm_vm->vcpu_refs[vcpu_handle], 1);
+
+	return vcpu_handle;
+}
+
+static int pkvm_arch_vcpu_create(struct pkvm_vcpu *pkvm_vcpu, struct fpstate *fps)
+{
+	struct kvm_vcpu *vcpu = &pkvm_vcpu->vcpu;
+
+	vcpu->kvm = &pkvm_vcpu->pkvm_vm->kvm;
+	/* Set cpu to -1 to indicate it is not loaded on any CPU */
+	vcpu->cpu = -1;
+	vcpu->vcpu_id = pkvm_vcpu->shared_vcpu->vcpu_id;
+
+	vcpu->arch.apic_base = pkvm_vcpu->shared_vcpu->arch.apic_base;
+
+	vcpu->arch.last_vmentry_cpu = -1;
+	vcpu->arch.regs_avail = ~0;
+	vcpu->arch.regs_dirty = ~0;
+	vcpu->arch.pat = MSR_IA32_CR_PAT_DEFAULT;
+
+	vcpu->arch.mce_banks = (void *)pkvm_vcpu + PKVM_VCPU_BASE_SIZE + kvm_vcpu_sz;
+	vcpu->arch.mci_ctl2_banks = (void *)vcpu->arch.mce_banks +
+				    KVM_MCE_SIZE;
+	vcpu->arch.mcg_cap = KVM_MAX_MCE_BANKS;
+
+	vcpu->arch.guest_fpu.fpstate = fps;
+
+	return kvm_x86_call(vcpu_create)(vcpu);
+}
+
+static void pkvm_unshare_vcpu(struct pkvm_vcpu *pkvm_vcpu)
+{
+	void *apic_regs = kern_pkvm_va(pkvm_vcpu->vcpu.arch.apic->regs);
+
+	pkvm_host_unshare_hyp(__pkvm_pa(pkvm_vcpu->shared_vcpu), kvm_vcpu_sz);
+	pkvm_host_unshare_hyp(__pkvm_pa(pkvm_vcpu->vcpu.arch.apic),
```

Comment:
As discussed in https://android-review.googlesource.com/c/kernel/common/+/3783094, apic may be legitimately NULL.